sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/GroupByKeyLoadTest.java

```diff
@@ -83,15 +83,14 @@ private GroupByKeyLoadTest(String[] args) throws IOException {
   void loadTest() throws IOException {
     Optional<SyntheticStep> syntheticStep = createStep(options.getStepOptions());
 
-    PCollection<KV<byte[], byte[]>> input =
-        pipeline.apply(SyntheticBoundedIO.readFrom(sourceOptions));
+    PCollection<KV<byte[], byte[]>> input = pipeline
+        .apply(SyntheticBoundedIO.readFrom(sourceOptions))
+        .apply(ParDo.of(new MetricsMonitor(METRICS_NAMESPACE)));
```

Comment:
The `MetricsMonitor` class creates the metrics. It's a parDo so everytime an element appears in a pipeline, it updates the desired metric. This pipeline does some "heavy-lifting" after the `MetricsMonitor` pardo is invoked. Doesn't this affect the results? Shouldn't we add the monitor at the very end of the pipeline?

Reply:
The issue appears to be that the pipeline was creating many `MetricsMonitor` steps based on the `options.getFanout()` for-loop. According to the [javadoc](https://beam.apache.org/releases/javadoc/2.8.0/org/apache/beam/sdk/metrics/Metrics.html), metrics are scoped to the applied transform they are executed in:

> Reported metrics are implicitly scoped to the transform within the pipeline that reported them. This allows reporting the same metric name in multiple places and identifying the value each transform reported, as well as aggregating the metric across

I'm not familiar with this pipeline or what it's testing. If moving the `MetricsMonitor` earlier in the pipeline retains correctness, this seems like the simplest solution. If you need the `MetricMonitor` after the fanout logic and aggregated, other options are:

1. Query the set of `MetricResult`s for each fork of the graph and aggregate them together locally.
1. Flatten the result of each flattened branch together and apply a single `MetricsMonitor` step to the flattened result.

Reply:
Thanks for the ideas. I considered the ideas before but wasn't sure that it is necessary to do it this way. It seems so. 

I think that for collecting the total pipeline run_time (which is a distribution metric) it's enough to flatten all the results of all distributions and get min and max to calculate it. Preferably it should be placed at the end of the pipeline to have all the processing time captured. 

For counting total bytes: it depends on a place where I measure it. It can be desired to have different sizes at the beginning and at the end of the pipeline. 

It will probably require splitting the `MetricsMonitor` to `TimeMonitor` and `BytesMonitor`. Time monitor can be applied anywhere in the pipeline (not much difference because we are looking for max and min time in the whole pipeline). Separate `BytesMonitor`s will calculate different results depending on the place in the pipeline they are "attached".

I will change this in some next contributions, now I wanted to show my thoughts. If you see flaws feel free to protest. :)

Reply:
I just took a look at the current [MetricsMonitor](https://github.com/apache/beam/blob/8a88e72f293ef7f9be6c872aa0dda681458c7ca5/sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/metrics/MetricsMonitor.java#L27) to get a better understanding of the intent.

The `runtime` counter will provide stats on the `DoFn` execution start/end, but that is only a subset of Pipeline execution. If you are trying to get pipeline start/stop time, you'll need to inject some counter calculation at the root(s) of the execution graph, and the sink(s). 

If you only care about Dataflow, you could consume some of the existing system counters being calculated automatically. Some details available at: [Using the Cloud Dataflow Monitoring Interface: Total Execution Time](https://cloud.google.com/dataflow/pipelines/dataflow-monitoring-intf#total-execution-time).

But, I believe these loadtests are run for many runners, so using Dataflow-native counters is likely not an option. In the Portability SDKs, we plan to have some counters reported by the SDK to all runners. /cc @ajamato

Reply:
Seems like this metric is being created as a UserMetric. Since this is the case, I see no issue with this change, and using it in whatever load tests we want.

We don't have any state sampling in the beam Java SDK yet, but the python one does. I am planing to add state sampling support to properly time step execution soon. Provides an estimate of time spend running in various steps of the pipeline, without incurring the time of expensive system calls (As in your use of System.currentTimeMillis).

This would provide estimates of the time (not perfectly accurate), but not add additional slowdown to the pipeline (which your System.currentTimeMillis calls are adding)